feat: support serverEntrypoint for self adapters (#15460)
withastro/astro: .changeset/fifty-dingos-study.md

---
'astro': minor
---

> Updates the Adapter API to allow providing a `serverEntrypoint` when using `entryType: 'self'`

Reviewer: ```suggestion
Adds a new `serverEntrypoint` to the Adapter API for specifying a custom entry point for an adapter
```
What about something that's a bit more use-case based here? The specifics of `entryType` maybe aren't super relevant for the opening sentence?

Author: I'm not sure. FYI `serverEntrypoint` is already available with `entryType: 'legacy-dynamic'` so this PR really is about supporting it when using `entryType: 'self'`. But maybe there's a better way to frame it, eg. allow doing X no matter the kind of adapter

Reviewer: I know there are more improvements to come anyway! This line isn't wrong, so I do think it's fine!

Astro 6 introduced a new powerful yet simple Adapter API for defining custom server entrypoints. You can now call `setAdapter()` with the new `entryType: 'self'` option and specify your custom `serverEntrypoint`:

The server entrypoint can now be provided as `serverEntrypoint` when calling `setAdapter()`:

```diff
export function myAdapter() {
    return {
        name: 'my-adapter',
        hooks: {
-            'astro:config:setup': ({ updateConfig }) => {
-                updateConfig({
-                    vite: {
-                        build: {
-                            rollupOptions: {
-                                input: 'my-adapter/server.js'
-                            }
-                        }
-                    }
-                })
-            },
            'astro:config:done': ({ setAdapter }) => {
                setAdapter({
                    name:'my-adapter',
                    entryType: 'self',
+                    serverEntrypoint: 'my-adapter/server.js',
                    supportedAstroFeatures: {
                        // ...
                    }
                })
            }
        }
    }
}

> ```

Author: ``````suggestion
```js
export function myAdapter() {
    return {
        name: 'my-adapter',
        hooks: {
            'astro:config:done': ({ setAdapter }) => {
                setAdapter({
                    name:'my-adapter',
                    entryType: 'self',
                    serverEntrypoint: 'my-adapter/server.js',
                    supportedAstroFeatures: {
                        // ...
                    }
                })
            }
        }
    }
}
```
``````

> If `serverEntrypoint` not set, Astro still expects a Rollup input.

Reviewer: Is there a good reason to provide a Rollup option *instead* of using this method? This seems way simpler (happy path!). If it's, e.g. Rollup still works for backwards compatibility reasons so no one's project breaks, but there's no real advantage to doing it, I might not even bother to mention that possibility. (And, I'd remove it from docs honestly.) If there *is* a good reason to choose that method, then probably a statement like:

```suggestion
For backwards compatibility (with previous beta versions of this API???), or to take advantage of further customizing `vite.build.rollupOptions`, you may instead specify your custom server entrypoint with a Rollup input.
```

Reviewer: I like it! But now we have to make sure that `vite.build.rollupOptions` isn't just coming in out of the blue from nowhere.

Does this make you want to update the earlier paragraph, something like...

```
Astro 6 introduced a new powerful yet simple Adapter API for defining custom server entrypoints, which previously required configuring Vite's Rollup options. You can now call `setAdapter()` with the `entryType: 'self'` option and specify your custom `serverEntrypoint`:
```

Or, add additional context here like...

```suggestion
If you need further customization at the Vite level, you can omit `serverEntrypoint` and instead specify your custom server entrypoint with [`vite.build.rollupOptions`](https://vite.dev/config/build-options#build-rollupoptions).
```
